src/components/tl/ucp/alltoall/alltoall_bruck.c

```diff
@@ -107,18 +110,38 @@ void ucc_tl_ucp_alltoall_bruck_progress(ucc_coll_task_t *coll_task)
     ucc_rank_t         tsize      = UCC_TL_TEAM_SIZE(team);
     ucc_coll_args_t   *args       = &TASK_ARGS(task);
     void              *scratch    = task->alltoall_bruck.scratch_mc_header->addr;
-    void              *mergebuf   = args->dst.info.buffer;
+    void              *mergebuf   = task->alltoall_bruck.dst;
     const ucc_rank_t   nrecv_segs = tsize / 2;
     const size_t       seg_size   = ucc_dt_size(args->src.info.datatype) *
                                     args->src.info.count / tsize;
     void              *data;
+    ucc_memory_type_t  smtype     = args->src.info.mem_type;
+    ucc_memory_type_t  dmtype     = args->dst.info.mem_type;
     ucc_rank_t sendto, recvfrom, step, index;
     ucc_rank_t level, snd_count;
     int send_buffer_index;
     ucc_status_t st;
+    ucc_ee_executor_t *exec;
+    ucc_ee_executor_task_args_t eargs;
+
+    if (task->alltoall_bruck.etask != NULL) {
+        st = ucc_ee_executor_task_test(task->alltoall_bruck.etask);
+        if (st == UCC_OK) {
+            ucc_ee_executor_task_finalize(task->alltoall_bruck.etask);
+            task->alltoall_bruck.etask = NULL;
+        } else {
+            if (ucc_unlikely(st < 0)) {
+                task->super.status = st;
+            }
+            return;
+        }
+    }
 
     if (task->alltoall_bruck.phase == PHASE_SENDRECV) {
         goto ALLTOALL_BRUCK_PHASE_SENDRECV;
+    } else if (task->alltoall_bruck.phase == PHASE_BCOPY) {
```

Comment:
It's only a suggestion, and a matter of taste, but I thought we usually implement this kind of dispatching like: 
```
switch(task->alltoall_bruck.phase) {
case PHASE_MERGE:
...
case PHASE_SENDRECV:
...
case PHASE_BCOPY:
...
}
```
I think it makes the flow a bit more readable
Maybe also it would be more readable to have `PHASE_MERGE` written explicitly here
But please do as you think is best